docs/RFCS/data_structures_for_logical_planning.md

```diff
@@ -0,0 +1,1001 @@
+types we'll need for them:
+
+- `(literal "N")` to represent the literal value N.
+- `(+ x y)` where `x` and `y` are indexes into the memo.
+- `(= x y)` where `x` and `y` are indexes into the memo.
+- `(filter x y)` where `x` is an index into the memo that designates
+  where the filter is getting data from, and `y` is an index into the
+  memo that designates the predicate being computed for each row to
+  decide whether the row is omitted.
+- `(project x y)` where `x` is an index into the memo that designates
+  where the projection is getting data from, and `y` is an index into
+  the memo that designates the value being computed for each row.
+- `(var "name")`, which represents the value in column `name` in
+  the current row in the relational context where it is evaluated.
+- `(scan <name> x)` which extracts all the rows in the named table into
+  the var(s) indexed by `x`.
+- `(exists x)` which evaluates to non-zero iff the result of the evaluation
+  of the expression indexed by `x` has more than zero rows.
+
+### Examples
+
+Let's take an first simple example:
+
+```sql
+select x+42 from foo
+```
+
+Corresponding memo, with properties omitted:
+
+```
+root index: 0
+0: (project 1 3)
+1: (scan "foo" 2)
+2: (var "x")
+3: (+ 2 4)
+4: (literal "42")
+```
+
+We can conceptually visualize this memo as an expression tree by
+drawing the edges between index references and the representations;
+there are some non-tree additional edges between variable definitions
+and uses:
+
+```
+         (project)
+          /     \
+  (scan "foo")    (+)
+       |        /   \
+    (var "x")--'   (lit 42)
+```
+
+Now, for a slightly more complex query with correlation:
+
+```sql
+select x from foo where exists (select a from t where a = x)
+```
+
+Memo, with properties omitted:
+
+```
+root index: 0
+0: (filter 1 3)
+1: (scan "foo" 2)
+2: (var "x")
+3: (exists 4)
+4: (filter 5 7)
+5: (scan "t" 6)
+6: (var "a")
+7: (= 2 6)
+```
+
+Or represented as a conceptual expression tree with additional
+non-tree edges between var def and use:
+
+```
+            (filter)
+             /    \
+     .------'    (exists)
+     |               \
+     |              (filter)
+     |             /       \
+ (scan "foo")  (scan "t")   '--.
+     |            |            |
+  (var "x")    (var "a")      (=)
+     \            \          /  |
+      \            '--------'   |
+       '------------------------'
+```
+
+## Early advantages
+
+Supposing we can use a single Go type to encode all the various
+expression representations (this is made possible by the [ADT
+RFC](20170517_algebraic_data_types.md)), the memo can be highly
+compact in memory and keep the number of allocations much lower than
+currently.
+
+### Where the magic happens, part 1: conjunctions
+
+Suppose we also consider AND expressions and (in)equality
+comparisons. For this we need the new m-expression type `(and z)`
+appears. `z` is an *integer set*, which we can conveniently encode as
```

Comment:
I've had to reverse engineer `z`. Perhaps explain that the integers in the set are indexes of (compatible) m-expressions of which `and (z)` represents the conjunction.